sklearn/cluster/tests/test_affinity_propagation.py

```diff
@@ -66,7 +67,14 @@ def test_affinity_propagation_predict():
     labels2 = af.predict(X)
     assert_array_equal(labels, labels2)
 
-
+    
+def test_affinity_propagation_convergence_warning():
+    # Test warning in AffinityPropagation.fit
+    af = AffinityPropagation(affinity="euclidean", max_iter=1)
+    assert_warns(ConvergenceWarning, af.fit, X)
+    assert_raises(ValueError, af.predict, X)
+    
```

Comment:
For example, low damping over a range of (outrageous low) preference values: 

	import numpy as np
	from sklearn.cluster import AffinityPropagation
	from sklearn.datasets.samples_generator import make_blobs
	from sklearn.metrics import euclidean_distances

	centers = np.array([[1, 1], [-1, -1], [1, -1]]) + 10
	X, _ = make_blobs(n_samples=60, n_features=2, centers=centers,
					  cluster_std=0.4, shuffle=True, random_state=0)

	S_sq = euclidean_distances(X, squared=True)
	S_nonsq = euclidean_distances(X, squared=False)

	max_sq = int(np.max(S_sq))
	max_nonsq = int(np.max(S_nonsq))

	# low damping over range of preference values

	for p in range(max_nonsq*10, max_nonsq*10+50):
		af = AffinityPropagation(preference=-p, max_iter=1000,
								 convergence_iter=100, damping=0.5,
								 affinity='precomputed')
		af.fit(-S_nonsq)
		n_clusters = len(af.cluster_centers_indices_) if af.cluster_centers_indices_ is not None else 0
		n_iter = af.n_iter_
		
		print 'p=%d: %d after %d iter' % (-p, n_clusters, n_iter)

	for p in range(max_sq*100, max_sq*100+50):
		af = AffinityPropagation(preference=-p, max_iter=1000,
								 convergence_iter=100, damping=0.5,
								 affinity='precomputed')
		af.fit(-S_sq)
		n_clusters = len(af.cluster_centers_indices_) if af.cluster_centers_indices_ is not None else 0
		n_iter = af.n_iter_
		
		print 'p=%d: %d after %d iter' % (-p, n_clusters, n_iter)

and you will get something like 

>p=-40: 2 after 231 iter
>p=-41: 1 after 1000 iter
>p=-42: 2 after 540 iter
>p=-43: 1 after 987 iter
>p=-44: 1 after 561 iter
>p=-45: 1 after 450 iter
>p=-46: 1 after 822 iter
>p=-47: 1 after 517 iter
>p=-48: 1 after 353 iter
>p=-49: 60 after 1000 iter
>p=-50: 2 after 337 iter
>p=-51: 0 after 1000 iter
>p=-52: 60 after 1000 iter
>p=-53: 60 after 1000 iter
>p=-54: 2 after 1000 iter
>p=-55: 1 after 988 iter
>p=-56: 60 after 1000 iter
>p=-57: 60 after 1000 iter
>p=-58: 59 after 1000 iter
>p=-59: 0 after 1000 iter
>p=-2000: 0 after 1000 iter
>p=-2001: 60 after 1000 iter
>p=-2002: 60 after 1000 iter
>p=-2003: 60 after 1000 iter
>p=-2004: 0 after 1000 iter
>p=-2005: 60 after 1000 iter
>p=-2006: 0 after 1000 iter
>p=-2007: 60 after 1000 iter
>p=-2008: 0 after 1000 iter
>p=-2009: 60 after 1000 iter
>p=-2010: 60 after 1000 iter
>p=-2011: 0 after 1000 iter
>p=-2012: 0 after 1000 iter
>p=-2013: 60 after 1000 iter
>p=-2014: 60 after 1000 iter
>p=-2015: 0 after 1000 iter
>p=-2016: 60 after 1000 iter
>p=-2017: 0 after 1000 iter
>p=-2018: 60 after 1000 iter
>p=-2019: 0 after 1000 iter

I suppose a test would look for IF # clusters = 0, THEN should have raised a ConvergenceWarning but does it make sense to let AP get away with return "60 clusters"  (clearly not converging as well ...)
Perhaps the real issue is that sklearn should include some sort of [preference range checking](http://www.psi.toronto.edu/affinitypropagation/preferenceRange.m)?